ui/src/utility/url-builder.spec.ts

```diff
@@ -0,0 +1,120 @@
+import {
+  objectToQuery,
+  queryToObject,
+  resolveApiUrl,
+  trimTrailingSlashes
+} from './url-builder';
+
+const queryObject = {
+  hello: 'my friend',
+  next: ['one', 'two', 'three'],
+  count: 2
+};
+
+const queryString =
+  '?hello=my%20friend&next[]=one&next[]=two&next[]=three&count=2';
+
+describe('Url Builder Tests', () => {
+  test('Test Complete Query to Object', () => {
+    expect(queryToObject(queryString)).toMatchObject(queryObject);
+  });
+
+  test('Test Complete Object to Query', () => {
+    expect(objectToQuery(queryObject)).toBe(queryString);
+  });
+
+  test('Test Empty Query', () => {
+    expect(queryToObject()).toMatchObject({});
+    expect(queryToObject('')).toMatchObject({});
+    expect(queryToObject(null)).toMatchObject({});
+    expect(queryToObject(undefined)).toMatchObject({});
+  });
+
+  test('Test Empty Object', () => {
+    expect(objectToQuery()).toBe('?');
+    expect(objectToQuery({})).toBe('?');
+    expect(objectToQuery(null)).toBe('?');
+    expect(objectToQuery(undefined)).toBe('?');
+  });
+
+  test('Test URL decoding', () => {
+    expect(
+      queryToObject(
+        "?ugly=()(*%26%5E%25%24%23%23%23%23%40!%3F%3E%3C%3A%60%22'%3A%3B%7B%7D%7C"
+      )
+    ).toMatchObject({
+      ugly: `()(*&^%$####@!?><:\`"':;{}|`
+    });
+  });
+
+  test('Test Object encoding', () => {
+    expect(
+      objectToQuery({
+        ugly: `()(*&^%$####@!?><:\`"':;{}|`
+      })
+    ).toBe(
+      "?ugly=()(*%26%5E%25%24%23%23%23%23%40!%3F%3E%3C%3A%60%22'%3A%3B%7B%7D%7C"
+    );
+  });
+});
+
+describe('Test Base URL Cleanup', () => {
+  test('Test strips off extra trailing slashes from the end', () => {
+    expect(trimTrailingSlashes('http://test.com///')).toBe('http://test.com');
+    expect(trimTrailingSlashes('http://test.com//')).toBe('http://test.com');
+    expect(trimTrailingSlashes('http://test.com')).toBe('http://test.com');
+    expect(trimTrailingSlashes('http://test.com/host/test/')).toBe(
+      'http://test.com/host/test'
+    );
+  });
+});
+
+describe('Resolves Expected URL', () => {
+  const locationMock = {
+    ancestorOrigins: <DOMStringList>{},
+    hash: '',
+    host: '',
+    href: '',
+    port: '',
+    search: '',
+    assign: jest.fn(),
+    reload: jest.fn(),
+    replace: jest.fn()
+  };
+
+  test('Resolves localhost', () => {
```

Comment:
@simpsont-oci I think these scenarios check out. Please confirm.

Reply:
Hmm, I'd have to test it, but I think this will only work for beetle when I'm in the office? If I've got an ssh tunnel on my home machine set up for beetle's http server, it seems like it's going to try to go the localhost route and query against `/bench2`? Or maybe I'm misunderstanding the code? I suppose I could always use 127.0.0.1 or my hostname and not have that problem?